server/src/main/java/org/opensearch/cluster/routing/allocation/decider/RemoteStoreMigrationAllocationDecider.java

```diff
@@ -0,0 +1,170 @@
+ * - For "REMOTE_STORE" migration direction:
+ *      - New primary shards can only be allocated to a remote node
+ *      - New replica shards can be allocated to a remote node iff the primary has been migrated/allocated to a remote node
+ * - For other directions ("DOCREP", "NONE"), the decision is always YES
+ *
+ * @opensearch.internal
+ */
+public class RemoteStoreMigrationAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "remote_store_migration";
+
+    private Direction migrationDirection;
+    private CompatibilityMode compatibilityMode;
+    private boolean remoteStoreBackedIndex;
+
+    public RemoteStoreMigrationAllocationDecider(Settings settings, ClusterSettings clusterSettings) {
+        this.migrationDirection = RemoteStoreNodeService.MIGRATION_DIRECTION_SETTING.get(settings);
+        this.compatibilityMode = RemoteStoreNodeService.REMOTE_STORE_COMPATIBILITY_MODE_SETTING.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(RemoteStoreNodeService.MIGRATION_DIRECTION_SETTING, this::setMigrationDirection);
+        clusterSettings.addSettingsUpdateConsumer(
+            RemoteStoreNodeService.REMOTE_STORE_COMPATIBILITY_MODE_SETTING,
+            this::setCompatibilityMode
+        );
+    }
+
+    private void setMigrationDirection(Direction migrationDirection) {
+        this.migrationDirection = migrationDirection;
+    }
+
+    private void setCompatibilityMode(CompatibilityMode compatibilityMode) {
+        this.compatibilityMode = compatibilityMode;
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        DiscoveryNode targetNode = node.node();
+
+        if (compatibilityMode.equals(CompatibilityMode.STRICT)) {
+            // assuming all nodes are of the same type (all remote or all non-remote)
+            return allocation.decision(
+                Decision.YES,
+                NAME,
+                getDecisionDetails(true, shardRouting, targetNode, " for strict compatibility mode")
+            );
+        }
+
+        if (migrationDirection.equals(Direction.REMOTE_STORE) == false) {
+            // docrep migration direction is currently not supported
+            return allocation.decision(
+                Decision.YES,
+                NAME,
+                getDecisionDetails(true, shardRouting, targetNode, " for non remote_store direction")
+            );
+        }
+
+        // check for remote store backed indices
+        IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
+        if (IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.exists(indexMetadata.getSettings())) {
+            remoteStoreBackedIndex = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(indexMetadata.getSettings());
+        }
+        if (remoteStoreBackedIndex && targetNode.isRemoteStoreNode() == false) {
```

Comment:
What about the case when the shard is replica here and primary has not been assigned to remote store. Will this block stop it's allocation to non-remote store node? Should we allow it? @gbbafna @ltaragi

Reply:
>What about the case when the shard is replica here and primary has not been assigned to remote store 

For a remote backed index . primary would always be assigned to remote store , followed by replica. A replica cannot be assigned before a primary shard is assigned .